worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala

```diff
@@ -1246,6 +1204,62 @@ class PushDataHandler(val workerSource: WorkerSource) extends BaseMessageHandler
     }
   }
 
+  private def writeLocalData(
+      fileWriters: Seq[FileWriter],
+      body: ByteBuf,
+      shuffleKey: String,
+      isPrimary: Boolean,
+      batchOffsets: Option[Array[Int]],
+      writePromise: Promise[Unit]): Unit = {
+    def writeData(fileWriter: FileWriter, body: ByteBuf, shuffleKey: String): Unit = {
+      try {
+        fileWriter.write(body)
+      } catch {
+        case e: Exception =>
+          if (e.isInstanceOf[AlreadyClosedException]) {
+            val (mapId, attemptId) = getMapAttempt(body)
+            val endedAttempt =
+              if (shuffleMapperAttempts.containsKey(shuffleKey)) {
+                shuffleMapperAttempts.get(shuffleKey).get(mapId)
+              } else -1
+            // TODO just info log for ended attempt
+            logWarning(s"Append data failed for task(shuffle $shuffleKey, map $mapId, attempt" +
+              s" $attemptId), caused by AlreadyClosedException, endedAttempt $endedAttempt, error message: ${e.getMessage}")
+          } else {
+            logError("Exception encountered when write.", e)
+          }
+          if (!writePromise.isCompleted) {
+            val cause =
+              if (isPrimary) {
+                StatusCode.PUSH_DATA_WRITE_FAIL_PRIMARY
+              } else {
+                StatusCode.PUSH_DATA_WRITE_FAIL_REPLICA
+              }
+            writePromise.failure(new CelebornIOException(cause))
+          }
+          fileWriter.decrementPendingWrites()
+      }
+    }
+    batchOffsets match {
+      case Some(batchOffsets) =>
+        batchOffsets.zip(fileWriters).foreach { case (offset, writer) =>
+          val length =
+            if (offset == batchOffsets.last) {
+              body.readableBytes() - offset
+            } else {
+              batchOffsets(batchOffsets.indexOf(offset) + 1) - offset
```

Comment:
I think the original `while` has better performance because it avoids calling `batchOffsets.indexOf (which is O(n) complexity)`, see https://github.com/apache/incubator-celeborn/pull/2086#issuecomment-1804944666

Reply:
sounds good